### PR TITLE
[GC] Add ZUnloadClassesFrequency for ZGC

### PR DESCRIPTION
**Pull request by tanghaoth90**

### CODE REVIEW

File: src/hotspot/share/gc/z/zBarrierSet.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ class ZBarrierSetC2;
 
 static BarrierSetNMethod* make_barrier_set_nmethod() {
   // NMethod barriers are only used when class unloading is enabled
-  if (!ClassUnloading) {
+  if (!ZHeap::heap()->should_unload_class()) {
```
Comment (mmyxym) [inline src/hotspot/share/gc/z/zBarrierSet.cpp:45]: Here should be 
(ClassUnloading && ZUnloadClassesFrequency != 0)

File: src/hotspot/share/gc/z/z_globals.hpp [hunk 2]
```diff
@@ -83,6 +83,10 @@
           "(can be any even integer between 4M and 32M)")                   \
           range(4*M, 32*M)                                                  \
                                                                             \
+  experimental(uint, ZUnloadClassesFrequency, 100,                          \
+          "Unload the classes every Nth ZGC cycle."                         \
+          "Set to zero to disable class unloading.")                        \
```
Comment (mmyxym) [inline src/hotspot/share/gc/z/z_globals.hpp:88]: Add constraint 0 - max_int?

Comment (tanghaoth90) [reply src/hotspot/share/gc/z/z_globals.hpp:88]: `java -XX:+UseZGC -XX:ZUnloadClassesFrequency=-1 -Xmx10g Xlog:gc*:gc.log:time`

Improperly specified VM option 'ZUnloadClassesFrequency=-1'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

no need to check if it is a negative integer.

File: src/hotspot/share/gc/z/zBarrierSet.cpp (hunk 1 above)
Comment (tanghaoth90) [reply src/hotspot/share/gc/z/zBarrierSet.cpp:45]: `ZUnloadClassesFrequency != 0` is in `ZHeap::should_unload_class()`

**Comment by tanghaoth90** [comment]
```
diff --git a/src/hotspot/share/gc/z/zBarrierSet.cpp b/src/hotspot/share/gc/z/zBarrierSet.cpp
index e4213352d4f..129cc2f74ad 100644
--- a/src/hotspot/share/gc/z/zBarrierSet.cpp
+++ b/src/hotspot/share/gc/z/zBarrierSet.cpp
@@ -42,7 +42,7 @@ class ZBarrierSetC2;
 
 static BarrierSetNMethod* make_barrier_set_nmethod() {
   // NMethod barriers are only used when class unloading is enabled
-  if (!ZHeap::heap()->should_unload_class()) {
+  if (ClassUnloading && ZUnloadClassesFrequency != 0) {
     return NULL;
   }
 
diff --git a/src/hotspot/share/gc/z/zUnload.cpp b/src/hotspot/share/gc/z/zUnload.cpp
index d1437379cda..3ea3c9f5fd3 100644
--- a/src/hotspot/share/gc/z/zUnload.cpp
+++ b/src/hotspot/share/gc/z/zUnload.cpp
@@ -36,8 +36,6 @@
 #include "gc/z/zHeap.inline.hpp"
 #include "oops/access.inline.hpp"
 
-static const ZStatSubPhase ZSubPhaseConcurrentClassesUnload("Concurrent Classes Unload");
-
 class ZIsUnloadingOopClosure : public OopClosure {
 private:
   ZPhantomIsAliveObjectClosure _is_alive;
@@ -122,8 +120,6 @@ void ZUnload::unload() {
     return;
   }
 
-  ZStatTimer timer(ZSubPhaseConcurrentClassesUnload);
-
   // Unlink stale metadata and nmethods
   // Purge stale metadata and nmethods that were unlinked
   unlink_and_purge();
```
@mmyxym I made a little change.

**Comment by tanghaoth90** [comment]
diff --git a/src/hotspot/share/gc/z/zBarrierSet.cpp b/src/hotspot/share/gc/z/zBarrierSet.cpp
index 129cc2f74ad..1d5a11ab7eb 100644
--- a/src/hotspot/share/gc/z/zBarrierSet.cpp
+++ b/src/hotspot/share/gc/z/zBarrierSet.cpp
@@ -42,7 +42,7 @@ class ZBarrierSetC2;
 
 static BarrierSetNMethod* make_barrier_set_nmethod() {
   // NMethod barriers are only used when class unloading is enabled
-  if (ClassUnloading && ZUnloadClassesFrequency != 0) {
+  if (!ClassUnloading || ZUnloadClassesFrequency == 0) {
     return NULL;
   }

**Comment by tanghaoth90** [comment]
```
diff --git a/src/hotspot/share/gc/z/zBarrierSet.cpp b/src/hotspot/share/gc/z/zBarrierSet.cpp
index 129cc2f74ad..1d5a11ab7eb 100644
--- a/src/hotspot/share/gc/z/zBarrierSet.cpp
+++ b/src/hotspot/share/gc/z/zBarrierSet.cpp
@@ -42,7 +42,7 @@ class ZBarrierSetC2;
 
 static BarrierSetNMethod* make_barrier_set_nmethod() {
   // NMethod barriers are only used when class unloading is enabled
-  if (ClassUnloading && ZUnloadClassesFrequency != 0) {
+  if (!ClassUnloading || ZUnloadClassesFrequency == 0) {
     return NULL;
   }
```
@mmyxym more change

**Review by mmyxym — APPROVED** [review]

**Review by linade — APPROVED** [review]

**Pull request merged**